kornia/contrib/models/rt_detr/architecture/rtdetr_head.py

```diff
@@ -0,0 +1,194 @@
+    ):
+        super().__init__()
+        self.num_queries = num_queries
+        self.input_proj = nn.ModuleList()
+        for ch_in in in_channels:
+            self.input_proj.append(ConvNormAct(ch_in, hidden_dim, 1, act="none"))
+
+        self.decoder_layers = nn.ModuleList()
+        for _ in range(num_decoder_layers):
+            layer = RTDETRTransformerDecoderLayer(hidden_dim, num_heads, len(in_channels), num_decoder_points)
+            self.decoder_layers.append(layer)
+
+        self.denoising_class_embed = nn.Embedding(num_classes, hidden_dim)  # not used in evaluation
+
+        self.query_pos_head = MLP(4, hidden_dim * 2, hidden_dim, 2)
+
+        self.enc_output = nn.Sequential(nn.Linear(hidden_dim, hidden_dim), nn.LayerNorm(hidden_dim))
+        self.enc_score_head = nn.Linear(hidden_dim, num_classes)
+        self.enc_bbox_head = MLP(hidden_dim, hidden_dim, 4, 3)
+
+        self.dec_score_head = nn.ModuleList()
+        self.dec_bbox_head = nn.ModuleList()
+        for _ in range(num_decoder_layers):
+            self.dec_score_head.append(nn.Linear(hidden_dim, num_classes))
+            self.dec_bbox_head.append(MLP(hidden_dim, hidden_dim, 4, 3))
+
+    def forward(self, fmaps: list[Tensor]) -> tuple[Tensor, Tensor]:
+        N = fmaps[0].shape[0]
+        fmaps = [proj(fmap) for proj, fmap in zip(self.input_proj, fmaps)]
+        spatial_shapes = [fmap.shape[2:] for fmap in fmaps]
+
+        feats = [fmap.flatten(2).permute(0, 2, 1) for fmap in fmaps]  # (N, C, H, W) -> (N, H*W, C)
+        memory = concatenate(feats, 1)  # rename to match original impl
+
+        anchors, valid_mask = self.generate_anchors(spatial_shapes)
+        out_memory = torch.where(valid_mask, memory, 0)
+        out_memory = self.enc_output(out_memory)
+
+        enc_out_logits = self.enc_score_head(out_memory)
+        enc_out_bboxes = anchors + self.enc_bbox_head(out_memory)
+
+        # only consider class with highest score at each spatial location
+        _, topk_indices = enc_out_logits.max(-1)[0].topk(self.num_queries, 1)  # (N, num_queries)
+
+        # alternative
+        # topk_bboxes = output_bboxes.gather(1, topk_indices.unsqueeze(-1).expand(-1, -1, 4))
+        batch_indices = torch.arange(N, dtype=topk_indices.dtype, device=topk_indices.device).unsqueeze(1)
+        ref_points = enc_out_bboxes[batch_indices, topk_indices]
+        tgt = out_memory[batch_indices, topk_indices]
+
+        for decoder_layer, bbox_head in zip(self.decoder_layers, self.dec_bbox_head):
+            ref_points_sigmoid = ref_points.sigmoid()
+            query_pos_embed = self.query_pos_head(ref_points_sigmoid)
+            tgt = decoder_layer(tgt, ref_points_sigmoid.unsqueeze(2), memory, spatial_shapes, query_pos_embed)
+            ref_points = bbox_head(tgt) + ref_points
+        logits = self.dec_score_head[-1](tgt)  # in evaluation, only last score head is used
+
+        return ref_points.sigmoid(), logits
+
+    def generate_anchors(
```

Comment:
I see also this as a nice utility called `create_spatial_anchors` or similar /cc @ducha-aiki @johnnv1

Reply:
Or some part of it

Reply:
yes, this should be useful for almost any detection model

Reply:
Perhaps the grid_xy (center point (cx, cy)) can be refactored for re-use? I see that it is quite similar to the YuNet code in Kornia.
https://github.com/kornia/kornia/blob/61462128dce2d95f25dd6f87213d2c11115a4630/kornia/contrib/face_detection.py#L397-L415

The anchor width and height can vary a lot, depending on how the detection model designs them though. So I think anchor width and height is hard to be refactored.

Reply:
Possibly but I’d do it in a posterior PR

Reply:
That makes sense. I will add a TODO for now.